pkg/webhook/resources/setting/validator.go

```diff
@@ -167,6 +165,9 @@ func NewValidator(
 	validateSettingFuncs[settings.HTTPProxySettingName] = validator.validateHTTPProxy
 	validateSettingUpdateFuncs[settings.HTTPProxySettingName] = validator.validateUpdateHTTPProxy
 
+	validateSettingFuncs[settings.UpgradeConfigSettingName] = validator.validateUpgradeConfig
```

Comment:
https://github.com/harvester/harvester/issues/6670 highlights a scenario we may encounter:

Assume we have a single-node cluster and set `restoreVM` to true in the UpgradeConfig setting. Now, we're about to perform an upgrade:
1. Before the upgrade, vm1 is in the running state. The `storeVM` function adds vm1 to the upgrade annotation.
2. During the upgrade, vm1 is stopped.
3. Once the upgrade is successfully completed, restoreVM starts vm1, and vm1 is in the starting state.
4. If we start another upgrade immediately, storeVM might not add vm1 to the upgrade annotation since it's still in the starting state.

There are two ways to fix this:

1. Wait until all VMs in the annotation are in the running state inside the restoreVM method.
2. Add a validation in the upgrade webhook and verify when restoreVM set to true there is no VMs are in the starting state before initiate an upgrade.

I personally prefer option 2, as it's simpler to implement.

c.c. @bk201, @w13915984028, @starbops

Reply:
As you said, a quick fix is to add an annotation to represent the last actions, and webhook check this. It can protect both vm image purging and vm restoring, and user can also manually delete it to stop if there is any kind of bugs.

A formal solution is to put those actions under a new state (upgrade crd already has many states) with time-out mechnism.

Reply:
> A formal solution is to put those actions under a new state (upgrade crd already has many states) with time-out mechnism.

As @w13915984028 pointed out, let's address this in [issue #6670](https://github.com/harvester/harvester/issues/6670) and keep the current behavior unchanged for now. I won’t be adding any further enhancements in this PR to avoid unnecessary complexity. Currently, if a user in a single-node cluster wants to perform a series of upgrades with restoreVM set to true, they will need to wait for the VMs to fully start before proceeding with the next upgrade.